accessors/storage/storage_accessor.go

```diff
@@ -0,0 +1,193 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package storageaccessor
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	storageclient "github.com/GoogleCloudPlatform/spanner-migration-tool/accessors/clients/storage"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/constants"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/logger"
+	"google.golang.org/api/googleapi"
+)
+
+// The StorageAccessor provides methods that internally use a storage client.
+// Methods should only contain generic logic here that can be used by multiple workflows.
+type StorageAccessor interface {
+	// Create a GCS bucket with the given name in the input projectId and location. If ttl is > 0,
+	// also apply a delete lifecycle rule with the input ttl and prefixes. Set @ttl to 0 to skip creating lifecycle rules.
+	CreateGCSBucket(ctx context.Context, sc storageclient.StorageClient, bucketName, projectID, location string, ttl int64, matchesPrefix []string) error
+	// Applies the bucket lifecycle with delete rule. Only accepts the Age and prefix rule conditions as it is only used for the Datastream destination
+	// bucket currently.
+	ApplyBucketLifecycleDeleteRule(ctx context.Context, sc storageclient.StorageClient, bucketName string, matchesPrefix []string, ttl int64) error
+	// UploadLocalFileToGCS uploads a local file at @localFilePath to a gcs file path @filePath with name @fileName.
+	UploadLocalFileToGCS(ctx context.Context, sc storageclient.StorageClient, filePath, fileName, localFilePath string) error
+	// Uploads a gcs object to gs://@filePath/@fileName with @data as content.
+	WriteDataToGCS(ctx context.Context, sc storageclient.StorageClient, filePath, fileName, data string) error
+	// Read a Gcs file path and returns the contents as a string.
+	ReadGcsFile(ctx context.Context, sc storageclient.StorageClient, filePath string) (string, error)
+	// Read a local or gcs file path. Files starting with a 'gs://' are treated as GCS files.
+	ReadAnyFile(ctx context.Context, sc storageclient.StorageClient, filePath string) (string, error)
+}
+
+// This implements the StorageAccessor interface. This is the primary implementation that should be used in all places other than tests.
+type StorageAccessorImpl struct{}
+
+func (sa *StorageAccessorImpl) CreateGCSBucket(ctx context.Context, sc storageclient.StorageClient, bucketName, projectID, location string, ttl int64, matchesPrefix []string) error {
```

Comment:
should we a create or update request to pass which can be shared among fxns?

Reply:
Done